File: cloud/src/dt_cloud/extras.py
"""Index extras: a provenance sidecar beside a scan's index tiers, written by
``path-index`` for a fresh scan and by ``dt-cloud index-extras`` as the backfill
for archived generations.

- ``attr.tsv``: every attributing prefix → ``user  source  evidence`` —
                the provenance of inferred ownership, which the path index
                drops on the way to its ``usr`` column.

It is **sorted by key, one line per entry, with a block index**
(``<name>.idx.json``: the first key and byte offset of every ~64 KB block), so
the site fetches just the byte range covering a view's subtree, far past what a
worker isolate should parse whole.

Keys are index paths: ``<bucket>/<dir>/…`` (no ``gs://``, no trailing slash).
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING

from utz import err

if TYPE_CHECKING:
    import pandas as pd

ATTR_FILE = "attr.tsv"
BLOCK = 64 * 1024


def write_blocked(path: Path, lines: list[str], key_of=lambda line: line) -> dict[str, int]:
    """Write ``lines`` (already sorted by key) as one text file plus a block
    index ``<path>.idx.json`` = ``{"v":1, "n": lines, "size": bytes,
    "keys": [first key per block], "offsets": [byte offset per block]}``.
    Blocks are cut at line boundaries at or past every ``BLOCK`` bytes."""
    keys: list[str] = []
    offsets: list[int] = []
    off = 0
    next_cut = 0
    with path.open("wb") as f:
        for line in lines:
            if off >= next_cut:
                keys.append(key_of(line))
                offsets.append(off)
                next_cut = off + BLOCK
            b = (line + "\n").encode()
            f.write(b)
            off += len(b)
    idx = {"v": 1, "n": len(lines), "size": off, "keys": keys, "offsets": offsets}
    path.with_name(path.name + ".idx.json").write_text(json.dumps(idx, separators=(",", ":")) + "\n")
    return {"lines": len(lines), "bytes": off, "blocks": len(keys)}


def attr_map(pfx_df: "pd.DataFrame") -> dict[str, list]:
    """``key → [user, source, evidence]`` from the prefix-label frame
    (``prefix_labels``; ``evidence`` when the frame carries it)."""
    has_ev = "evidence" in pfx_df.columns
    out: dict[str, list] = {}
    for row in pfx_df.itertuples(index=False):
        ev = getattr(row, "evidence", None) if has_ev else None
        out[row.key] = [row.user, getattr(row, "source", None), None if ev is None or ev != ev else ev]
    return out


def write_extras(pfx_df: "pd.DataFrame | None", out_dir: Path) -> dict[str, int]:
    """Given a prefix frame, write ``attr.tsv`` (+ index) into ``out_dir``;
    returns entry counts. No frame → no sidecar (the reader treats an absent
    sidecar as "no extras for this generation")."""
    counts: dict[str, int] = {}
    if pfx_df is not None:
        out_dir.mkdir(parents=True, exist_ok=True)
        attr = attr_map(pfx_df)
        rows = [f"{k}\t{u}\t{src or ''}\t{ev or ''}" for k, (u, src, ev) in sorted(attr.items())]
        write_blocked(out_dir / ATTR_FILE, rows, key_of=lambda line: line.split("\t", 1)[0])
        counts["attr"] = len(attr)
    err(f"extras: {out_dir}: " + ", ".join(f"{k}={v:,}" for k, v in counts.items()))
    return counts

File: site/functions/_lib/extras.ts
/**
 * Index extras (specs/index-extras.md): per-scan sidecars beside the index
 * tiers — `ck.txt` (checkpoint-shaped dirs, decided over each dir's FULL
 * child list) and `attr.tsv` (every attributing prefix → user, source,
 * evidence). Both are sorted by key with a block index (`<name>.idx.json`),
 * so a view fetches only the byte range covering its subtree (keys under
 * `P` are contiguous in sorted order) plus, for provenance, the blocks
 * holding P's ancestors. The small indexes are memoized per (scan,
 * generation) per isolate; a scan without sidecars (older generations, until
 * `dt-cloud index-extras` backfills) yields `null` and every consumer keeps
 * today's behaviour.
 */
import type { Env } from './auth.js'
import { makeStore } from './index.js'
import { shared } from './shared.js'

export type Provenance = [source: string, evidence: string | null, prefix: string]

/** A dir whose own name says checkpoint (mirrors `dt_cloud.extras.CKPT_NAME_RE` and the client's `looksCkpt`). */
export const CKPT_NAME_RE = /(^|[-_.])(ckpts?|checkpoints?)([-_.]|$)/i

export interface BlockIndex { v: number; n: number; size: number; keys: string[]; offsets: number[] }

/** The most a view will pull of one sidecar: past this (a bucket root over a
 *  sidecar with hundreds of thousands of lines) the extras are skipped for
 *  that view rather than parsed whole. */
const MAX_RANGE = 8 * 1024 * 1024

export interface ExtrasView {
  /** Checkpoint-shaped dirs under (and at) the view root. */
  ck: ReadonlySet<string>
  /** The provenance of `user`'s bytes under `path`: the deepest attributing
   *  ancestor-or-self whose user is `user`, or null. */
  provenance(path: string, user: string): Provenance | null
}

const idxMemo = new Map<string, Promise<{ dir: string; ck: BlockIndex | null; attr: BlockIndex | null } | null>>()
const TTL_MS = 10 * 60_000
// A scan WITHOUT sidecars is re-probed after a minute, so a backfill shows
// up promptly rather than after the isolate's next recycle. Tracked by
// timestamp, not a timer: a `setTimeout` armed inside a request can be
// dropped once that request's context ends.
const MISS_TTL_MS = 60_000
const missAt = new Map<string, number>()

async function readJson(env: Env, key: string): Promise<unknown | null> {
  try {
    const { bytes } = await makeStore(env).get(key)
    return JSON.parse(new TextDecoder().decode(bytes))
  } catch {
    return null
  }
}

/** The scan's sidecar indexes (or null when the generation has none). */
async function indexes(env: Env, date: string) {
  if (!env.DB) return null
  const r = await env.DB.prepare('SELECT dir FROM index_schema WHERE date = ? AND variant = ?').bind(date, 'path').first<{ dir: string | null }>()
  const dir = r?.dir ?? `listing/${date}`
  const key = `${date}|${dir}`
  const miss = missAt.get(key)
  if (miss != null && Date.now() - miss > MISS_TTL_MS) { idxMemo.delete(key); missAt.delete(key) }
  const got = await shared(idxMemo, key, async () => {
    const [ck, attr] = await Promise.all([readJson(env, `${dir}/ck.txt.idx.json`), readJson(env, `${dir}/attr.tsv.idx.json`)])
    if (!ck && !attr) return null
    return { dir, ck: ck as BlockIndex | null, attr: attr as BlockIndex | null }
  }, TTL_MS)
  if (!got) missAt.set(key, missAt.get(key) ?? Date.now())
  else missAt.delete(key)
  return got
}

/** Whether the scan has sidecars — for cache keys, so a backfill isn't
 *  hidden behind responses cached before it landed. */
export async function hasExtras(env: Env, date: string): Promise<boolean> {
  return (await indexes(env, date)) != null
}

/** Index of the block whose first key ≤ `key` (−1 before the first). */
export function blockOf(idx: BlockIndex, key: string): number {
  let lo = 0
  let hi = idx.keys.length - 1
  let ans = -1
  while (lo <= hi) {
    const mid = (lo + hi) >> 1
    if (idx.keys[mid] <= key) { ans = mid; lo = mid + 1 } else hi = mid - 1
  }
  return ans
}

/** The byte span `[start, end)` covering blocks `from..to` (inclusive),
 *  clamped to the index; null when the range is empty — `to < from`, or
 *  `from` past the last block (an empty sidecar has zero blocks, so *every*
 *  range is empty). Guards the reader from fabricating a `[undefined, 0]`
 *  range out of an empty `offsets`, which becomes a malformed store GET. */
export function blockSpan(idx: BlockIndex, from: number, to: number): [number, number] | null {
  if (from < 0) from = 0
  if (to < from || from >= idx.offsets.length) return null
  const start = idx.offsets[from]
  const end = to + 1 < idx.offsets.length ? idx.offsets[to + 1] : idx.size
  return end <= start ? null : [start, end]
}

/** The lines of blocks `from..to` (inclusive), or null when over budget. */
async function readBlocks(env: Env, key: string, idx: BlockIndex, from: number, to: number): Promise<string[] | null> {
  const span = blockSpan(idx, from, to)
  if (!span) return []
  const [start, end] = span
  if (end - start > MAX_RANGE) return null
  const { bytes } = await makeStore(env).get(key, { offset: start, length: end - start })
  const text = new TextDecoder().decode(bytes)
  return text.split('\n').filter(Boolean)
}

/** Every key in `[lo, hi)`, via the block index. */
async function rangeLines(env: Env, key: string, idx: BlockIndex, lo: string, hi: string): Promise<string[] | null> {
  const a = blockOf(idx, lo)
  const b = blockOf(idx, hi)
  return readBlocks(env, key, idx, a, b < 0 ? 0 : b)
}

const ancestorsOf = (path: string): string[] => {
  const out: string[] = []
  for (let p = path; p; p = p.includes('/') ? p.slice(0, p.lastIndexOf('/')) : '') out.push(p)
  return out
}

/**
 * The extras a view rooted at `path` needs: the checkpoint dirs under it and
 * an attribution map covering its subtree plus its ancestors. Null when the
 * scan has no sidecars.
 */
export async function extrasFor(env: Env, date: string, path: string): Promise<ExtrasView | null> {
  const ix = await indexes(env, date)
  if (!ix) return null
  const lo = path
  const hi = path === '' ? '￿' : path + '0' // '0' sorts just past '/'
  const [ckLines, attrLines, ancLines] = await Promise.all([
    ix.ck ? rangeLines(env, `${ix.dir}/ck.txt`, ix.ck, lo, hi) : Promise.resolve([]),
    ix.attr ? rangeLines(env, `${ix.dir}/attr.tsv`, ix.attr, lo, hi) : Promise.resolve([]),
    // The root's ancestors: one block each (deduped), for inherited provenance.
    ix.attr
      ? Promise.all([...new Set(ancestorsOf(path).slice(1).map(a => blockOf(ix.attr!, a)))].map(b => readBlocks(env, `${ix.dir}/attr.tsv`, ix.attr!, b, b)))
      : Promise.resolve([]),
  ])
  return viewOf(ckLines ?? [], [...(attrLines ?? []), ...ancLines.flatMap(l => l ?? [])])
}

/** An `ExtrasView` over already-fetched lines (the pure half, for tests). */
export function viewOf(ckLines: string[], attrLines: string[]): ExtrasView {
  const ck = new Set(ckLines)
  const attr = new Map<string, [string, string | null, string | null]>()
  for (const line of attrLines) {
    const [k, u, src, ev] = line.split('\t')
    if (k && u) attr.set(k, [u, src || null, ev || null])
  }
  const provenance = (p: string, user: string): Provenance | null => {
    for (const a of ancestorsOf(p)) {
      const hit = attr.get(a)
      if (hit && hit[0] === user) return [hit[1] ?? 'unknown', hit[2], a]
    }
    return null
  }
  return { ck, provenance }
}
